core/src/main/scala/chisel3/Probe.scala

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3._
+import chisel3.Data.ProbeInfo
+import chisel3.reflect.DataMirror.isProbe
+import chisel3.internal.{requireIsChiselType, Builder}
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.sourceinfo.SourceInfo
+import chisel3.internal.firrtl._
+
+/** Utilities for creating and working with probe types.
+  */
+object Probe {
+
+  private def apply_impl[T <: Data](source: => T, writable: Boolean): T = {
+    val prevId = Builder.idGen.value
+    val data = source // should only evaluate source once
+    requireIsChiselType(data)
+    val ret = if (!data.mustClone(prevId)) data else data.cloneType.asInstanceOf[T]
+    ret.probeInfo = Some(ProbeInfo(writable))
+    ret
+  }
+
+  /** Create a read-only probe type from a Chisel type. This is only used for
+    * ports and not for hardware.
+    */
+  def apply[T <: Data](source: => T): T = apply_impl(source, false)
+
+  /** Create a writable probe type from a Chisel type. This is only used for
+    * ports and not for hardware.
+    */
+  def writable[T <: Data](source: => T): T = apply_impl(source, true)
+
+  private def requireIsProbe(probeExpr: Data): Unit = {
+    require(isProbe(probeExpr), s"expected $probeExpr to be a probe.")
+  }
+
+  private def requireIsWritableProbe(probeExpr: Data): Unit = {
+    requireIsProbe(probeExpr)
+    require(probeExpr.probeInfo.get.writable, s"expected $probeExpr to be writable.")
+  }
+
+  /** Initialize a Probe with a provided probe expression. */
+  def define(sink: Data, probeExpr: Data)(implicit sourceInfo: SourceInfo): Unit = {
+    requireIsProbe(sink)
+    requireIsProbe(probeExpr)
+    pushCommand(ProbeDefine(sourceInfo, sink.ref, probeExpr.ref))
+  }
+
+  private def probe_impl[T <: Data](source: => T, writable: Boolean): Data = {
+    val prevId = Builder.idGen.value
+    val t = source
+
+    // construct probe to return with cloned info
+    val clone = if (!t.mustClone(prevId)) t else t.cloneTypeFull
+    clone.bind(chisel3.internal.ProbeBinding(Builder.forcedUserModule, Builder.currentWhen, t))
+    clone.setRef(ProbeExpr(t.ref))
+    clone.probeInfo = Some(ProbeInfo(writable))
+
+    clone
+  }
+
+  /** Create a read-only probe. */
```

Comment:
how is this different than the Probe.apply?

Reply:
This is creating a probe value rather than probe type, as we discussed a few days ago. I've separated these out into their own objects, so hopefully the updated API is clearer!